### PR TITLE
Feat: add menu for tabs

### PR DESCRIPTION
**Pull request by kenyx89**
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| Fixed tickets    | https://github.com/marktext/marktext/issues/1369
| License          | MIT

### Description

Added menu for tabs.

--

### CODE REVIEW

File: src/renderer/contextMenu/tabs/index.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  const CONTEXT_ITEMS = [CLOSE_THIS, CLOSE_OTHERS, CLOSE_SAVED, CLOSE_ALL, SEPARATOR, RENAME, COPY_PATH, SHOW_IN_FOLDER]
+  const FILE_CONTEXT_ITEMS = [RENAME, COPY_PATH, SHOW_IN_FOLDER]
+
+  FILE_CONTEXT_ITEMS.forEach(item => {
+    item.enabled = !!pathname
+  })
+
+  CONTEXT_ITEMS.forEach(item => {
+    const menuItem = new MenuItem(item)
+    menuItem.tab = tab
```
Comment (fxha) [inline src/renderer/contextMenu/tabs/index.js]: This is not the best way to solve the problem but please only use tab id (`tab.id`). You can get the original tab by id in editor. Maybe a function call with the tab id to construct the item would be a better way.

File: src/renderer/components/editorWithTabs/tabs.vue [hunk 2]
```diff
@@ -82,8 +86,51 @@ export default {
...
+  created () {
+    this.$nextTick(() => {
+      // listen for bus events.
+      bus.$on('TABS::close-this', this.closeTab)
+      bus.$on('TABS::close-others', this.closeOthers)
+      bus.$on('TABS::close-saved', this.closeSaved)
+      bus.$on('TABS::close-all', this.closeAll)
+      bus.$on('TABS::rename', this.rename)
+      bus.$on('TABS::copy-path', this.copyPath)
+      bus.$on('TABS::show-in-folder', this.showInFolder)
```
[the remark this answers is not in the document]
Comment (Jocs) [reply src/renderer/components/editorWithTabs/tabs.vue]: @fxha We also use Upper case in sidebar, like bellow,

```js
bus.$on('SIDEBAR::show-in-folder', () => {
      const { pathname } = state.activeItem
      shell.showItemInFolder(pathname)
    })
```

So in the later stage we should unify the naming method, namespace + eventname, namespace need to be capitalized, eventname use snack style naming, 'SIDEBAR::foo-bar'.

File: src/renderer/store/editor.js [hunk 3]
```diff
@@ -681,6 +681,32 @@ const actions = {
     }
   },
 
+  CLOSE_OTHER_TABS ({ state, dispatch }, file) {
+    const { tabs } = state
+    tabs.filter(f => f.id !== file.id).forEach(tab => {
+      dispatch('CLOSE_TAB', tab)
```
Comment (Jocs) [inline src/renderer/store/editor.js]: We have `CLOSE_TABS` action, so no need to `dispatch('CLOSE_TAG', tab)` each time?

File: src/renderer/store/editor.js [hunk 4]
```diff
@@ -681,6 +681,32 @@ const actions = {
...
+    })
+  },
+
+  CLOSE_SAVED_TABS ({ state, dispatch }) {
+    const { tabs } = state
+    tabs.filter(f => f.isSaved).forEach(tab => {
+      dispatch('CLOSE_TAB', tab)
```
Comment (Jocs) [inline src/renderer/store/editor.js]: see above.

File: src/renderer/store/editor.js [hunk 5]
```diff
@@ -681,6 +681,32 @@ const actions = {
...
+    })
+  },
+
+  CLOSE_ALL_TABS ({ state, dispatch }) {
+    const { tabs } = state
+    tabs.slice().forEach(tab => {
+      dispatch('CLOSE_TAB', tab)
```
Comment (Jocs) [inline src/renderer/store/editor.js]: see above.

**Comment by Jocs** [comment]
@fxha @kenyx89 I recommend adding a menu item for the `new tab` and renaming `show in file manager` to `show in folder` so that each menu item is uniform in length.

File: src/renderer/store/editor.js (hunk 3 above)
Comment (kenyx89) [reply src/renderer/store/editor.js]: @Jocs `CLOSE_TABS` mutation forces to close all unsaved tabs without save prompt. So, if we dispatch `CLOSE_TABS` once, all unsaved tabs will be lost.

File: src/renderer/contextMenu/tabs/index.js (hunk 1 above)
Comment (Jocs) [reply src/renderer/contextMenu/tabs/index.js]: How about menuItem._tabId ? add a prefix `_`

File: src/renderer/components/editorWithTabs/tabs.vue [hunk 6]
```diff
@@ -82,13 +82,15 @@ export default {
       const newLeft = Math.max(0, Math.min(tabs.scrollLeft + delta, tabs.scrollWidth))
       tabs.scrollLeft = newLeft
     },
-    closeTab (tab) {
-      if (tab.id) {
+    closeTab (tabId) {
+      const tab = this.tabs.find(f => f.id === tabId)
```
Comment (Jocs) [inline src/renderer/components/editorWithTabs/tabs.vue]: Do not repeat yourself.

Comment (kenyx89) [reply src/renderer/components/editorWithTabs/tabs.vue]: Well... to be honest, I don't understand why @fxha didn't like to pass `tab` object itself to menu item.
@fxha Can I know your opinion?

Comment (fxha) [reply src/renderer/components/editorWithTabs/tabs.vue]: I think personal preferences but also that we don't spread object all over. All newer API/functions should use just id's and older need refactoring.  Electrons menu items are main process component and if we bind object to it, the values are copied between main and renderer process which is not necessary.

You can simply forward the tab id to the vue function if possible like:

```js
// file 1
this.$store.dispatch('CLOSE_OTHER_TABS', tabId)

// store/editor.js

CLOSE_OTHER_TABS ({ state, dispatch }, tabId) {
    const { tabs } = state
    tabs.filter(f => f.id !== tabId).forEach(tab => {
// ...
}
````

**Comment by fxha** [comment]
Thanks @kenyx89

**Pull request merged**